File: 111-二叉树的最小深度/111.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
/*
执行时间为8ms
使用DFS的思想进行搜索
当根节点为空时 直接返回0
当左子树为空时 直接返回右子树+1
当右子树为空时 直接返回左子树+1
*/

class Solution {
public:
    int minDepth(TreeNode* root) {
        if(!root) return 0;
        if(!root->right) return 1 + minDepth(root->left);
        if(!root->left) return 1 + minDepth(root->right);
        
        int min_left = minDepth(root->left);
        int min_right = minDepth(root->right);
        
        int result = 1 + min(min_left,min_right);
        
        return result;
    }
};

File: 111-二叉树的最小深度/111.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public int minDepth(TreeNode root) {
      if(root == null) return 0;
      
      else if(root.left == null){
          return minDepth(root.right) + 1;
      }
      else if(root.right == null){
          return minDepth(root.left) + 1;
      }
      else{
          return Math.min(minDepth(root.left), minDepth(root.right)) + 1;
      }
  }
}